### PR TITLE
Max7219 Clean Up

### PR DESCRIPTION
**Pull request by Roxy-3D**
Some of the Rotations did not do the right thing with setting and clearing pixels.
I think it is correct now.   But if not...  It is much closer to being correct.

### CODE REVIEW

File: Marlin/src/gcode/feature/leds/M7219.cpp [hunk 1]
```diff
@@ -71,12 +71,18 @@ void GcodeSuite::M7219() {
...
-      SERIAL_EOL();
+    SERIAL_ECHO("]=");
+    for (int8_t j = 7; j >= 0; j--) {
+      if ( LEDs[x] & (0x01<<j) )
+        SERIAL_ECHO("1");
+      else
+        SERIAL_ECHO("0");
     }
+    SERIAL_EOL();
     return;
```
Comment (thinkyhead) [inline Marlin/src/gcode/feature/leds/M7219.cpp:84]: This `return` is inside the `x` loop. I'll patch it in a moment…

File: Marlin/src/gcode/feature/leds/M7219.cpp [hunk 2]
```diff
@@ -71,12 +71,18 @@ void GcodeSuite::M7219() {
...
+    for (int8_t x = 0; x < 8 * MAX7219_NUMBER_UNITS; x++) {
       SERIAL_ECHOPAIR("LEDs[", x);
-      SERIAL_ECHOPAIR("]=", LEDs[x]);
-      SERIAL_EOL();
+    SERIAL_ECHO("]=");
+    for (int8_t j = 7; j >= 0; j--) {
+      if ( LEDs[x] & (0x01<<j) )
+        SERIAL_ECHO("1");
+      else
+        SERIAL_ECHO("0");
```
Comment (thinkyhead) [inline Marlin/src/gcode/feature/leds/M7219.cpp:81]: These are all slightly smaller code-wise. I'll pick one for the patch.
```cpp
SERIAL_ECHOPGM("1");
SERIAL_ECHO(int(TEST(LEDs[x], j)));
SERIAL_CHAR('0' + TEST(LEDs[x], j));
SERIAL_CHAR(TEST(LEDs[x], j) ? '1' : '0');
```

Comment (thinkyhead) [reply Marlin/src/gcode/feature/leds/M7219.cpp:81]: Hmm… `SERIAL_PRINT(LEDs[x], 2)` should print any `int` in binary format. I'll test that too…

File: Marlin/src/feature/Max7219_Debug_LEDs.h [hunk 3]
```diff
@@ -116,19 +116,19 @@ void Max7219_idle_tasks();
...
   #define MAX7219_Y_LEDS          8
   #define MAX7219_X_LEDS          (MAX7219_Y_LEDS * (MAX7219_NUMBER_UNITS))
-  #define XOR_7219(x, y)          LEDs[y + (MAX7219_X_LEDS - 1 - (x)) & 0xF8] ^= _BV((x & 0x07))
-  #define SET_PIXEL_7219(x, y)    LEDs[y + (MAX7219_X_LEDS - 1 - (x)) & 0xF8] |= _BV((x & 0x07))
-  #define CLEAR_PIXEL_7219(x, y)  LEDs[y + (MAX7219_X_LEDS - 1 - (x)) & 0xF8] &= (_BV((x & 0x07)) ^ 0xFF)
-  #define BIT_7219(x, y)          TEST(LEDs[y + (MAX7219_X_LEDS - 1 - (x)) & 0xF8], ((x & 0x07)))
+  #define XOR_7219(x, y)          LEDs[x + (y & 0xF8)] ^= _BV((x & 0x07))
+  #define SET_PIXEL_7219(x, y)    LEDs[x + (y & 0xF8)] |= _BV((x & 0x07))
+  #define CLEAR_PIXEL_7219(x, y)  LEDs[x + (y & 0xF8)] &= (_BV((x & 0x07)) ^ 0xFF)
+  #define BIT_7219(x, y)          TEST(LEDs[x + (y & 0xF8)], ((x & 0x07)))
```
Comment (thinkyhead) [inline Marlin/src/feature/Max7219_Debug_LEDs.h:131]: If the idea is just to have a single byte for each "row" of LEDs, then all we need is `LEDs[y]`. There's no need to do anything more with it. In the normal case, `y` gives the row and `x` gives the bit. In first the 90° rotated case `x` gives the array index, and `y` gives the bit. For the other two cases, we additionally reverse the array and/or bit index. (i.e., `LEDs[total_rows - row]` and/or `7 - col`).

The `& 0x7` on the bit index is just for safety, and not really needed.

Comment (thinkyhead) [reply Marlin/src/feature/Max7219_Debug_LEDs.h:131]: The assumptions I was making in my previous overhaul to Max7219 include…
- Hardware is at one abstraction level where "row" always means the byte and "column" always means the bit. So the `LEDs` array always reflects the expected hardware state.
- At the higher (user/gcode) level, `X` always refers to left-right and `Y` always refers to up-down. So when the display is rotated 90° or 180°, the `X` and `Y` indices switch roles.
- The index into the `LEDs` array is simply the hardware row. So, it should only be keyed to `X` or `Y`, and not both. Therefore `LEDs[x + (y & 0xF8)]` should be either `x`, `y`, `rows-x` or `rows-y` depending on orientation.

…and so on…

I have a PR pending and I'm in the process of testing it with two chained units right now…

**Comment by thinkyhead** [comment]
So when you chain your units together… Are you expecting the second one to show up to the RIGHT of the first one in the chain, or to the LEFT? Currently the code is treating the second unit as being to the left of the first one.

**Comment by Roxy-3D** [comment]
What I expected... (and tested)  is a chain of devices.    Any chain is always 8 x (n*8)    So, sometimes (For a rotation of 90 or -90)  X is only 8 but Y is more...   Or... (for a rotation of 0 or 180)  Y is 8 but X is more....      That is pretty well tested.

**Comment by Roxy-3D** [comment]
It is possible our Max7219 units are laid out different.   But I have 3 different flavors of them.   The ones I have look like this:   https://www.ebay.com/itm/MAX7219-LED-8x8-Dot-Matrix-Display-Module-MCU-control-New/272806646960?ssPageName=STRK%3AMEBIDX%3AIT&_trksid=p2057872.m2749.l2649

> If the idea is just to have a single byte for each "row" of LEDs

Yes...   But that isn't quite right....   The LED[] matrix just holds the bits for the display independent of how the display is orientated.    The macro definitions get the the right bits set or cleared within the matrix depending on orientation.     (It does get confusing!!!!)

The LED[0-7] is always the first display.   LED[8-15] is always the 2nd display.   and LED[16-23] is always the 3rd display.   (Does that help???)

Connect up 2 or 3 displays and see if it works....    It is possible you have displays that are different than mine.   But it kind of seems like there is a 'standard' on how to chain these gizmos together.

**Comment by thinkyhead** [comment]
> The LED[] matrix just holds the bits for the display independent of how the display is orientated. 

Correct. Which is why they are `uint8_t`.

**Comment by thinkyhead** [comment]
> The LED[] matrix just holds the bits for the display independent of how the display is orientated. 

Correct. Which is why they are `uint8_t`.

> The LED[0-7] is always the first display. LED[8-15] is always the 2nd display. and LED[16-23] is always the 3rd display.

That is what I assumed.

> seems like there is a 'standard' on how to chain these gizmos together.

Try testing yours with a `MAX7219_ROTATE` of `0` and see if it does what you expect with the current code. I've got a PR to submit shortly which just cleans up the code a bit and preserves the existing behavior.

**Comment by Roxy-3D** [comment]
> Try testing yours with a MAX7219_ROTATE of 0 and see if it does what you expect with the current code. I've got a PR to submit shortly which just cleans up the code a bit and preserves the existing behavior.

I've been playing darts (and winning way too many beers).    I'll try what ever ever is there in the morning!!!  :)     I'll try all four orientations with what ever code you merge in.

**Comment by Roxy-3D** [comment]
@thinkyhead   This message is duplicated just to make sure you see it...

At a -90 degree orientation with 2 units...   X=8  &  Y=16    The column command works OK.   I can do a `M7219 C2 V????`  and get the right 16 bit pattern drawn on the display.   However,  the row command is broken.   A `M7219 R2 V??` of any 8 bit value fails to work.   But the init_pattern=2 spiral pattern does look correct.  

At a 0 degree orientation with 2 units...   X=16  &  Y=8    The init pattern also seems to work correctly. 
And...  It appears both the `M7219 R ?? V????` and `M7219 C ?? V????` work correctly.

At a 90 degree orientation with 2 units...    X=8  &  Y=16   Init spiral is wrong!   `M7219 R?? V??` is totally scrambled.   `M7219 C??  V??` is backwards.   The LED's are sort of correct, but they are a mirror image.

At 180 degree orientation with 2 units...  X=16  &  Y=8.   Init pattern is sort of correct.   But the two halves need to be exchanged for it to be correct.    The LED's are in reverse order for `M7219 C2 V254`.     `M7219 C8 V255` puts the line on the wrong 8x8 matrix.

How about we revert everything and just do one change at a time.   Otherwise this is going to be too complicated to figure out.   If we make one change at a time...   We will know when we go from a working state to a broken state.   Perhaps the first thing to do is converting to a singleton with no logic changes.

**Comment by thinkyhead** [comment]
Do you find that the previous iteration of the code before the merge works properly in all those orientations?

**Comment by Roxy-3D** [comment]
The stuff I merged worked.   But I haven't been tracking or testing all the changes after that.
How about we revert the code and turn it into a singleton.   I know how to test that to make sure it works.

**Comment by thinkyhead** [comment]
I reverted the code, and it's behaving just the same as the current code, except for the obvious bugs. Let's just take a look at 0° and 90° to start with…

With a rotation of 90° the right edge of the second matrix is the "top." This is fine if the units are arranged left-to-right, but it's wrong if the second unit is placed to the "left" of the first, as shown here:

img_3993

From your previous descriptions it sounded like your units were chained right-to-left. So, can you re-confirm for me that the old code (0b2d686) works correctly with your Max7219 in all 4 orientations? That will be a good start, and tell me whether your units are logically arranged left-to-right or right-to-left.

I'll take care of the obvious bugs giving odd results once we're well settled on how they are supposed to be arranged.

**Comment by Roxy-3D** [comment]
Actually...  I think I know where the root of the problem is...    It is true my units are chained together right to left.     (The unit on the right is where the wires plug in.)       But I think those units you have are laid out different.    I think the LED matrix is rotated 90 degrees from where they are on mine.

So...  when you stack them next to each other, the individual LED's are all in different places.

I hesitate to suggest this...   But besides having a way to control the orientation of the entire string of LED matrix's.    I think we need a control to say how the 8x8 LED matrix is oriented relative to the plug and wires.   (I think your LED's are rotated 90 degrees from the plug wires compared to mine.)

Huuuuugggghhhhh!!!!!

I actually have one unit that looks just like yours.   I'll test it later this afternoon to see where it thinks (0,0) is compared to units I'm using.   And for that matter...  where (7,7) is too!

**UPDATE:    The units you have turned the LED matrix 180 degrees from the other 3 flavors I have.   With the connector on the right...  And an orientation of 0 degrees, (7,7) is in the upper left for your units.   On all 3 of the other styles (7,7) is in the lower right.    I think that explains the problem.    But also...  The units you have look like they aren't made to be chained in a contiguous stripe.   There is no way to butt the LED's of one matrix up against the other ones next to it if you chain straight from D-out of one unit to D-in of the next unit.  The wires need to criss cross on your units.**

**Comment by thinkyhead** [comment]
Mine have the connector on the left.

**Comment by thinkyhead** [comment]
In native orientation mine have the connector on the left. So register 1 is the top row here, and register 8 is the bottom row.
image

> ~~I think the LED matrix is rotated 90 degrees from where they are on mine.~~
~~So when you set register 1 to 255 it fills the first column on the left?~~

> The units you have turned the LED matrix 180 degrees from the other 3 flavors I have. 

So your plug is on the right when register 1 is at the top…

> The units you have look like they aren't made to be chained in a contiguous stripe. 

True. They would have to be arranged side-by-side (chip at the bottom) to line up well. But I'm trying to use an arrangement that matches an all-in-one (like yours) while testing.

> I think we need a control to say how the 8x8 LED matrix is orientated relitive to the plug and wires.

I'm sure it's achievable. Just have to get to know the common arrangements.

> And an orientation of 0 degrees, (7,7) is in the upper left for your units.

In other words, the native orientation has the plug on the left. It's shown in the video:
https://youtu.be/JXMEKXsmTpA?t=1m40s

**Comment by Roxy-3D** [comment]
You have your unit flipped 180 degrees from the 0 degree orientation.    For testing, I have your unit with the connector on the right and set to 0 degrees.     And the same thing for the other 3 types of displays.

Doing a M7215 R0 V240  lights up the right 4 LED's on the bottom row on your display.
Doing a M7215 R0 V240  lights up the left 4 LED's on the top row on my displays.

**Comment by thinkyhead** [comment]
Please post some images of the units you're using, running the 1.1.9 code, and indicate on each one how the coordinates go. Please show all 4 orientations. I will set up my units to act like yours and get it set up in that way. After I get that solid I'll look into how best to chain my individual units.

**Comment by thinkyhead** [comment]
> Doing a `M7215 R0 V240` lights up the right 4 LED's on the bottom row on your display.

Well, no. The plug belongs on the left on my displays. That's their native orientation.

I'm working with the basic assumption that zero rotation means the top row is register 1. So these must have the wires on the left in the "native" or "zero rotation" orientation.

**Comment by Roxy-3D** [comment]
> I'm working with the basic assumption that zero rotation means the top row is register 1.

OK...   And doing that will make the first unit show up right.   But if you chain them...   The next unit is on the wrong side of the first unit. 

I tried to dump the LED[] matrix, and the M7219 P no longer works...   
```
  if (parser.seen('P')) {
    for (uint8_t r = 0; r < MAX7219_ROWS; r++) {
      SERIAL_ECHOPGM("led_line[");
      if (r < 10) SERIAL_CHAR('_');
      SERIAL_ECHO(r);
      SERIAL_ECHO("]=");
      for (uint8_t b = 8; b--;) SERIAL_CHAR('0' + TEST(max7219.led_line[r], b));
      SERIAL_EOL();
    }
  }
```
It would seem to me the b loop should be starting at bit 7 ??????    Or perhaps, the TEST(max7219.led_line[r], b-1)     ---  Then the loop would terminate correctly.
But some how MAX7219_ROWS is ending up at 0.

**Comment by thinkyhead** [comment]
> OK... And doing that will make the first unit show up right. But if you chain them... The next unit is on the wrong side of the first unit.

In my case the next unit ends up to the right of the first. If the multi-matrix units arrange them from right-to-left then I can finagle mine into that orientation in order to match what you see.

**Comment by Roxy-3D** [comment]
OK!   So I think what this means is we just need one simple option...   If it mirrors the bits going out during a send() everything will match up.    I'm thinking send() should still do the same thing for each display module...  But for your display type it needs to update each module in the reverse order.

**Comment by thinkyhead** [comment]
Yes, for a more complete solution we will need to specify whether the displays are arranged left-to-right or right-to-left when chained.

In the case of the all-in-one units it sounds like the first row sent goes to the rightmost matrix. But… maybe it goes out to the leftmost? The datasheet says the first register sent ends up going to the last unit in the chain (if more than one register is sent, otherwise it only goes to the first unit).

As far as implementation details go, I'll ensure that the `LEDs` array is set so it always matches up to the natural arrangement of the hardware. That way it can be clocked out exactly as it is. Then, the orientation will **only** affect how the bits in the array are set.

**Comment by Roxy-3D** [comment]
> Yes, for a more complete solution we will need to specify whether the displays are arranged left-to-right or right-to-left when chained.

> In the case of the all-in-one units the first row sent goes to the rightmost matrix. With these individual units, they natively chain to the right.

I guess it depends...   On the back of mine, the silk screen text is correct...  And the connector for data in is on the right.    But the important thing is...   The data for the furthest out chained unit is sent first.  (when looking at the back of the board.)

> As far as implementation details go, I'll ensure that the LEDs array is set so it always matches up to the natural arrangement of the hardware. That way it can be clocked out exactly as it is. Then, the orientation will only affect how the bits in the array are set.

OK...   I think this means we can rotate an individual display's bits, right?   But that is independent from whether the left most or right most display gets is data first or last, right?

**Comment by thinkyhead** [comment]
> OK... I think this means we can rotate an individual display's bits, right?

That could be implemented, but for now the idea is to simply know the orientation of the whole thing so we know what XY mean when setting bits in the `LEDs` array.

> But that is independent from whether the left most or right most display gets is data first or last, right?

To get perfect separation between the hardware and the abstract level, it's simplest to have the `LEDs` array always clock out exactly the same way. For a two-unit matrix: 0, 8; 1, 9; 2, 10; 3, 11; 4, 12; 5, 13; 6, 14; 7, 15. For three-unit, 0, 8, 16; 1, 9, 17; etc.

Then, when setting a row, pixel, column, etc. we just set the bits in the array according to the specified rotation and arrangement.

**Comment by thinkyhead** [comment]
Oops, let me revise. The clock-out order should be 8, 0; 9, 1; … so the bits for the last unit get clocked out first.

**Comment by Roxy-3D** [comment]
Yeah....    For me, I prefer the high order bits are on the left (for a row) and top (for a column).
But as long as it is consistent across all displays...   It really doesn't matter.     

I guess this means we should change having the connector on the left, with units chained to the right is an orientation of 0 degrees.

**Comment by thinkyhead** [comment]
Yeah, the native orientation of these units is to have register 1 at the top, and bits naturally arranged with high bit 7 on the left. That way you can send a simple bitmap in forward order and it won't be flipped horizontally or upside-down.

For example:
```gcode
M7219 R0 V24    ; ...**...
M7219 R1 V60    ; ..****..
M7219 R2 V126   ; .******.
M7219 R3 V219   ; **.**.**
M7219 R4 V255   ; ********
M7219 R5 V90    ; .*.**.*.
M7219 R6 V129   ; *......*
M7219 R7 V66    ; .*....*.
```

**Comment by thinkyhead** [comment]
Ok, so #11596 should do the trick. I tested all the orientations and they all seem good.

G-code I used for testing: [Max7219-Orient.gcode.txt](https://github.com/MarlinFirmware/Marlin/files/2301110/Max7219-Orient.gcode.txt)
